### ISSUE TITLE
6.16 ctcsound issue

### ISSUE THREAD

**Issue by enchilada404** [csound/csound#1502]
I'm on win 10 64bit.

When I try to import ctcsound, this happens:

```
Python 3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license()" for more information.
>>> import ctcsound
Traceback (most recent call last):
  File "<pyshell#0>", line 1, in <module>
    import ctcsound
  File "C:\Program Files\Csound6_x64\bin\ctcsound.py", line 2699, in <module>
    libcspt = ct.CDLL(ctypes.util.find_library("csnd6"))
  File "C:\Program Files\Python39\lib\ctypes\__init__.py", line 364, in __init__
    if '/' in name or '\\' in name:
TypeError: argument of type 'NoneType' is not iterable 
```

It's working fine in 6.15

**Comment by fggp** [comment]
Nothing has changed within ctcsound.py between 6.15 and 6.16 about loading
libraries in windows systems. The only difference I see is about the Python
version. Which Python version did you use with 6.15 and which Python
version do you use with 6.16?

François


Le mer. 14 juil. 2021 à 06:31, enchilada404 ***@***.***> a
écrit :

**Comment by enchilada404** [comment]
I have Python 3.8 and 3.9 installed, both gives the same error. I have not tried with any earlier Python version for 6.16. But I did try uninstalling 6.16, reinstalling 6.15, and ctcsound imports like it should (in Python 3.9):

```
Python 3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license()" for more information.
>>> import ctcsound
>>> 
```

**Comment by fggp** [comment]
So it seems that with 6.16 the library csnd6 is not found. I'm sorry, but I
don't have a windows box to test it right now. I'll have a try tomorrow if
I can install csound on a machine running window10...

François

Le jeu. 15 juil. 2021 à 09:44, enchilada404 ***@***.***> a
écrit :

**Comment by vlazzarini** [comment]
I see this code

```
elif sys.platform.startswith('win'):
    if sys.version_info.major <=3 and sys.version_info.minor < 8:
        libcspt = ct.cdll.csnd6
    else:
        libcspt = ct.CDLL(ctypes.util.find_library("csnd6"))
```

so if Python < 3.8 then it uses ` ct.cdll.csnd6`.  I wonder if it works in Python 3.7. It is really

**Comment by vlazzarini** [comment]
Also this Python bug might be relevant

https://bugs.python.org/issue39243

**Comment by fggp** [comment]
This code had been added, because there was an issue with Python 3.8 on
Windows. It seems that this issue was only happening with Python 3.8. I
just made an installation on a Windows 10 (64-bit) machine with Csound 6.16
and Python 3.9.6 and I got this error message:

Traceback (most recent call last):
  File "<pyshell#6>", line 1, in <module>
    import ctcsound
  File
"C:\Users\francois.pinot\AppData\Local\Programs\Python\Python39\lib\ctcsound.py",
line 2699, in <module>
    libcspt = ct.CDLL(ctypes.util.find_library("csnd6"))
  File
"C:\Users\francois.pinot\AppData\Local\Programs\Python\Python39\lib\ctypes\__init__.py",
line 364, in __init__
    if '/' in name or '\\' in name:
TypeError: argument of type 'NoneType' is not iterable

Doing the change below in ctcsound.py fixes the problem:

elif sys.platform.startswith('win'):
    if sys.version_info.major==3 and sys.version_info.minor==8:
        libcsound = ct.CDLL(ctypes.util.find_library("csound64"))
    else:
        libcsound = ct.cdll.csound64

But then another problem arises:

FileNotFoundError: Could not find module 'csnd6' (or one of its
dependencies). Try using the full path with constructor syntax.

It seems that the Csound 6.16 Windows installer lacks csnd6.dll. This
library is needed by ctcsound to bind with CsPerfThread...

François

Le jeu. 15 juil. 2021 à 17:59, vlazzarini ***@***.***> a
écrit :

**Comment by vlazzarini** [comment]
I checked yesterday and the csnd6.dll is present in the zip binaries package. Not sure why it's not in the installer. Could you see if that csnd6.dll is usable?

@skyne would you be able to check why the installer is not including csnd6.dll? Thanks

Prof. Victor Lazzarini
Maynooth University
Ireland

On Jul 16, 2021, at 9:52 AM, fggp ***@***.***> wrote:

﻿
*Warning*

This email originated from outside of Maynooth University's Mail System. Do not reply, click links or open attachments unless you recognise the sender and know the content is safe.

This code had been added, because there was an issue with Python 3.8 on
Windows. It seems that this issue was only happening with Python 3.8. I
just made an installation on a Windows 10 (64-bit) machine with Csound 6.16
and Python 3.9.6 and I got this error message:

Traceback (most recent call last):
File "<pyshell#6>", line 1, in <module>
import ctcsound
File
"C:\Users\francois.pinot\AppData\Local\Programs\Python\Python39\lib\ctcsound.py",
line 2699, in <module>
libcspt = ct.CDLL(ctypes.util.find_library("csnd6"))
File
"C:\Users\francois.pinot\AppData\Local\Programs\Python\Python39\lib\ctypes\__init__.py",
line 364, in __init__
if '/' in name or '\\' in name:
TypeError: argument of type 'NoneType' is not iterable

Doing the change below in ctcsound.py fixes the problem:

elif sys.platform.startswith('win'):
if sys.version_info.major==3 and sys.version_info.minor==8:
libcsound = ct.CDLL(ctypes.util.find_library("csound64"))
else:
libcsound = ct.cdll.csound64

But then another problem arises:

FileNotFoundError: Could not find module 'csnd6' (or one of its
dependencies). Try using the full path with constructor syntax.

It seems that the Csound 6.16 Windows installer lacks csnd6.dll. This
library is needed by ctcsound to bind with CsPerfThread...

François

Le jeu. 15 juil. 2021 à 17:59, vlazzarini ***@***.***> a
écrit :

**Comment by fggp** [comment]
Well, I will have a try on monday (I don't have Windows at home...). But
I'm quite sure it should work.

François

Le ven. 16 juil. 2021 à 15:44, vlazzarini ***@***.***> a
écrit :

> I checked yesterday and the csnd6.dll is present in the zip binaries
> package. Not sure why it's not in the installer. Could you see if that
> csnd6.dll is usable?
>
> @skyne would you be able to check why the installer is not including
> csnd6.dll? Thanks
>
> Prof. Victor Lazzarini
> Maynooth University
> Ireland
>

**Comment by vlazzarini** [comment]
If csnd6.dll is not there, I think CsoundQT won't work either, and Blue. This is very strange.

**Comment by vlazzarini** [comment]
So I checked this and yes, no csnd6.dll in the installer, it's my own fault, I removed it by mistake when getting rid of the csnd6 Python module. I've restored it on develop, and as soon as I know it's building properly, I'll make a new installer. For the moment, @enchilada404 , I would suggest you to download the zip for 6.16.1 and get the csnd6.dll from there and place it alongside Csound64.dll. That should get you running with Python 3.9.
